[DSM] DDP-7556_Mercury_fixes

pegahtah opened the pull request:
PR comments applied and changes to the request to also have collaborator participant id

pegahtah committed:
DDP-7556
* added the PR comments
* made the request to work with collaborator pt id as well

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_fixes

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/ddp/instance/DDPInstanceDao.java:
```diff
@@ -42,18 +42,21 @@ public class DDPInstanceDao implements Dao<DDPInstanceDto> {
     private static final String SQL_SELECT_INSTANCE_WITH_ROLE =
             "SELECT ddp_instance_id, instance_name, base_url, collaborator_id_prefix, migrated_ddp, billing_reference, "
-                    + "es_participant_index, es_activity_definition_index, es_users_index, study_pre_filter, research_project, "
+                    +
```
can we not do just `+` in a line… move your `mercury….` into the same line as `(select`, please

pegahtah replied:
ugh I missed this , I'll fix

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java:
```diff
@@ -98,44 +96,28 @@ public void onSuccess(String messageId) {
         String json = new Gson().toJson(mercuryPdoOrder);
         if (StringUtils.isNotBlank(json)) {
-            orderOnMercury(projectId, topicId, json, barcodes, ddpParticipantId, mercuryOrderId);
+            this.mercuryOrderDao.orderOnMercury(projectId, topicId, json, barcodes, ddpParticipantId, mercuryOrderId, this);
```
I don't like to give in `this` to `orderOnMercury` please refactor to have the dao method return what is needed and then order it from here!

pegahtah replied:
So if you look at the `OrderOnMercury` I am doing the insertion of db and the publish on pub sub in thee same method to be  able to roll back the db when sth goes wrong with pubsub.

SimoneMaiwald commented:
please also remove `mercury_order_creator` from `instance_settings` on dev if you haven't done so yet

pegahtah committed:
DDP-7556
* fixed the checkstyle error

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_fixes

# Conflicts:
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/DDPInstance.java

pegahtah committed:
DDP-7556
* moved try catch into `createAndPublishMessage` method and removed it from Dao

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_fixes

SimoneMaiwald commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java:
```diff
@@ -98,44 +95,34 @@ public void onSuccess(String messageId) {
+
     }
+
+    private boolean requestHadCollaboratorId(String ddpParticipantId, String collaboratorParticipantId) {
```
is that just for testing purpose? we should know what they send in the message

pegahtah committed:
DDP-7556
* moved the `createAllOrders` before the pubsub publishing to check for existing

pegahtah committed:
DDP-7556
* added the ability to always have test tissues with sample id

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_fixes

pegahtah committed:
Applying PR comments and fixes to the branch
*added mercuryOrderCreator to DDPInstanceDao
*fixed the SQL query for getting eligible tissues and samples
*fixed the json to have all the info in a `productOrder` object
*changed published json on pubsub to be base64
*added userId and ddpInstanceId columns to the backend to keep better track of the orders

pegahtah committed:
Merge remote-tracking branch 'origin/develop' into DDP-7556_Mercury_fixes

# Conflicts:
#	pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/CreateClinicalDummyKitRoute.java

pegahtah committed:
Applying PR comments and fixes to the branch
*fixed merge conflicts

pegahtah committed:
Applying PR comments and fixes to the branch
*renamed Route to`PostMercuryOrderDummyRoute`
* removed `ddpParticipant Id` from `MercuryOrderDummyRequest`
* renamed `MercuryOrderRequest` class to `MercuryOrderDummyRequest`

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/mercury/PostMercuryOrderDummyRoute.java:
```diff
@@ -0,0 +1,67 @@
+import spark.Route;
+
+@Slf4j
+public class PostMercuryOrderDummyRoute implements Route {
```
renamed route, added `Dummy` in it

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/statics/DBConstants.java:
```diff
@@ -512,7 +513,5 @@ public class DBConstants {
     public static final String COHORT_TAG_NAME = "cohort_tag_name";
 
-    //instance_settings
-    public static final String MERCURY_ORDER_CREATOR = "mercury_order_creator";
```
moved to ddp_instance

pegahtah commented on pepper-apis/dsm-core/src/main/resources/liquibase/CMI/DDP-7556_mercury_sequencing.xml:
```diff
@@ -26,12 +26,20 @@
     </changeSet>
+    <changeSet id="DDP-7556_additional_mercury_sequencing" author="pegah">
+        <addColumn tableName="mercury_sequencing">
+            <column name="created_by" type="VARCHAR(200)"></column>
```
user who submits the order

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/route/CreateClinicalDummyKitRoute.java:
```diff
@@ -96,13 +95,20 @@ public Object handle(Request request, Response response) {
-            if (maybeParticipantByParticipantId.isEmpty()) {
-                throw new RuntimeException("PT not found " + ddpParticipantId);
+            while (tries < 10 && (maybeParticipantByParticipantId.isEmpty() || !participantIsEnrolled(maybeParticipantByParticipantId)
+                    || maybeParticipantByParticipantId.get().getProfile().map(ESProfile::getHruid).isEmpty())) {
```
I moved the participant enrolled check into this method

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java:
```diff
@@ -98,44 +101,31 @@ public void onSuccess(String messageId) {
         String mercuryOrderId = createMercuryUniqueOrderId();
         MercuryPdoOrder mercuryPdoOrder = new MercuryPdoOrder(creatorId, mercuryOrderId, researchProject, barcodes);
-        String json = new Gson().toJson(mercuryPdoOrder);
+        MercuryPdoOrderBase mercuryPdoOrderBase = new MercuryPdoOrderBase(mercuryPdoOrder);
```
changes in Json needed all the order going in to one `productOrder` object

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java:
```diff
@@ -98,44 +101,31 @@ public void onSuccess(String messageId) {
-                    }
-                }
-                publishWithErrorHandler(projectId, topicId, json);
+                List<MercuryOrderDto> newOrders = MercuryOrderUseCase.createAllOrders(barcodes, ddpParticipantId, mercuryOrderId, userId);
```
Creating all the `MercuryOrderDto` objects for submitted barcodes

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/pubsub/MercuryOrderPublisher.java:
```diff
@@ -98,44 +101,31 @@ public void onSuccess(String messageId) {
-                publishWithErrorHandler(projectId, topicId, json);
+                List<MercuryOrderDto> newOrders = MercuryOrderUseCase.createAllOrders(barcodes, ddpParticipantId, mercuryOrderId, userId);
+                this.publishWithErrorHandler(projectId, topicId, json);
+                this.mercuryOrderDao.insertMercuryOrders(newOrders);
```
only inserting if the pubsub message was successfully published

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercuryOrderDao.java:
```diff
@@ -17,43 +19,49 @@
-                stmt.setString(2, barcode);
-                stmt.setString(3, ddpParticipantId);
+            try (PreparedStatement stmt = conn.prepareStatement(SQL_GET_KIT_FROM_BARCODE_KIT_LABEL)) {
+                stmt.setString(1, ddpParticipantId);
```
finding all eligible kit labels for participant

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dao/mercury/MercuryOrderDao.java:
```diff
@@ -63,9 +71,33 @@ public Optional<MercuryOrderDto> getMercuryOrderFromSMIdOrKitLabel(String barcod
+        results = inTransaction((conn) -> {
+            SimpleResult dbVals = new SimpleResult();
+            try (PreparedStatement stmt = conn.prepareStatement(SQL_GET_KIT_FROM_BARCODE_SM_ID)) {
+                stmt.setString(1, ddpParticipantId);
```
finding all eligible sm ids for participant

pegahtah commented on pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/db/dto/mercury/MercuryOrderUseCase.java:
```diff
@@ -0,0 +1,32 @@
+            orders.add(order);
+        });
+        return orders;
+    }
```
Finding all the kits and sm ids that are eligible for ordering for that participant in here, choosing the ones that were in the order and adding them to the orders object.

If a label was not eligible, throwing an error

pegahtah committed:
Merge branch 'develop' into DDP-7556_Mercury_fixes

pegahtah committed:
Merge branch 'develop' into DDP-7556_Mercury_fixes
